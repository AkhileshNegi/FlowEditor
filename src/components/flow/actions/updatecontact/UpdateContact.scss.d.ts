// This file is automatically generated by the "typings-for-css-modules" Webpack loader.
// https://github.com/Jimdo/typings-for-css-modules-loader

export const value: string;
