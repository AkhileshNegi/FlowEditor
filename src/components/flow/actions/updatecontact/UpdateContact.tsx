import * as React from 'react';
import { Types } from '~/config/interfaces';
import {
    SetContactAttribute,
    SetContactChannel,
    SetContactLanguage,
    SetContactName
} from '~/flowTypes';
import { emphasize } from '~/utils';

const withEmph = (text: string, emph: boolean) => (emph ? emphasize(text) : text);

export const renderSetText = (
    name: string,
    value: string,
    emphasizeName: boolean = false
): JSX.Element => {
    if (value) {
        return (
            <div>
                Set {withEmph(name, emphasizeName)} to {emphasize(value)}.
            </div>
        );
    } else {
        return <div>Clear {withEmph(name, emphasizeName)}.</div>;
    }
};

const UpdateContactComp: React.SFC<SetContactAttribute> = action => {
    switch (action.type) {
        case Types.set_contact_field:
            return renderSetText(action.field.name, action.value, true);
        case Types.set_contact_channel:
            return renderSetText('channel', (action as SetContactChannel).channel.name);
        case Types.set_contact_language:
            return renderSetText('language', (action as SetContactLanguage).language);
        case Types.set_contact_name:
            return renderSetText('name', (action as SetContactName).name);
    }
};

export default UpdateContactComp;
