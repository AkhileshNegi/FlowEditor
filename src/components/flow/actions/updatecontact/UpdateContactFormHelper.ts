import {
    assetToChannel,
    assetToField,
    assetToLanguage,
    channelToAsset,
    fieldToAsset,
    languageToAsset
} from '~/components/flow/actions/updatecontact/helpers';
import {
    CHANNEL_PROPERTY,
    LANGUAGE_PROPERTY,
    NAME_PROPERTY,
    UpdateContactFormState
} from '~/components/flow/actions/updatecontact/UpdateContactForm';
import { FormHelper, getTypeConfig, Types } from '~/config/typeConfigs';
import {
    SetContactAttribute,
    SetContactChannel,
    SetContactField,
    SetContactLanguage,
    SetContactName
} from '~/flowTypes';
import { NodeEditorSettings } from '~/store/nodeEditor';

export class UpdateContactFormHelper implements FormHelper {
    public initializeForm(settings: NodeEditorSettings): UpdateContactFormState {
        const state: UpdateContactFormState = {
            type: Types.set_contact_name,
            valid: false,
            name: { value: '' },
            channel: { value: null },
            language: { value: null },
            field: { value: NAME_PROPERTY },
            fieldValue: { value: '' }
        };

        if (settings.originalAction) {
            const originalType = settings.originalAction.type;
            // these have aliases, so compare the config we resolve to
            if (getTypeConfig(originalType) === getTypeConfig(Types.set_contact_field)) {
                state.type = originalType;

                switch (originalType) {
                    case Types.set_contact_field:
                        const fieldAction = settings.originalAction as SetContactField;
                        state.field = { value: fieldToAsset(fieldAction.field) };
                        state.fieldValue = { value: fieldAction.value };
                        state.valid = true;
                        return state;
                    case Types.set_contact_channel:
                        const channelAction = settings.originalAction as SetContactChannel;
                        state.field = { value: CHANNEL_PROPERTY };
                        state.channel = { value: channelToAsset(channelAction.channel) };
                        state.valid = true;
                        return state;
                    case Types.set_contact_language:
                        const languageAction = settings.originalAction as SetContactLanguage;
                        state.field = { value: LANGUAGE_PROPERTY };
                        state.valid = true;
                        state.language = {
                            value: languageToAsset({
                                iso: languageAction.language,
                                name: languageAction.language
                            })
                        };
                        return state;
                    case Types.set_contact_name:
                        const nameAction = settings.originalAction as SetContactName;
                        state.valid = true;
                        state.name = {
                            value: nameAction.name
                        };
                        return state;
                }
            }
        }

        // default is updating name
        return state;
    }

    public stateToAction(uuid: string, state: UpdateContactFormState): SetContactAttribute {
        if (state.type === Types.set_contact_field) {
            return {
                uuid,
                type: state.type,
                field: assetToField(state.field.value),
                value: state.fieldValue.value
            };
        }

        if (state.type === Types.set_contact_channel) {
            return {
                uuid,
                type: state.type,
                channel: assetToChannel(state.channel.value)
            };
        }

        if (state.type === Types.set_contact_language) {
            return {
                uuid,
                type: state.type,
                language: assetToLanguage(state.language.value)
            };
        }

        if (state.type === Types.set_contact_name) {
            return {
                uuid,
                type: state.type,
                name: state.name.value
            };
        }
    }
}
