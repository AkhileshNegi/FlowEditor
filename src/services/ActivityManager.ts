import Counter from '~/components/counter/Counter';
import { Exit } from '~/flowTypes';

// tslint:disable:ban-types
// how often we ask the server for new data
const REFRESH_SECONDS = 10;

/**
 * Contains all the activity data for a flow
 */
export interface Activity {
    // exit_uuid:destination_uuid -> count
    segments: { [key: string]: number };

    // node_uuid -> count
    nodes: { [key: string]: number };
}

export default class ActivityManager {
    private static singleton: ActivityManager;

    // our main activity fetch from the external
    private activity: Activity;

    // our simulation activity
    private simulation: Activity;

    private flowUUID: string;

    private getActivityExternal: Function;

    private listeners: { [key: string]: Counter } = {};
    private timer: any;

    constructor(flowUUID: string, getActivity: Function) {
        this.flowUUID = flowUUID;
        this.getActivityExternal = getActivity;

        this.clearSimulation = this.clearSimulation.bind(this);
        this.setSimulation = this.setSimulation.bind(this);
        this.fetchActivity = this.fetchActivity.bind(this);
        this.notifyListeners = this.notifyListeners.bind(this);
        this.deregister = this.deregister.bind(this);
        this.registerListener = this.registerListener.bind(this);
        this.getActivity = this.getActivity.bind(this);
        this.getActiveCount = this.getActiveCount.bind(this);
        this.getPathCount = this.getPathCount.bind(this);

        this.fetchActivity();
    }

    public clearSimulation(): void {
        this.simulation = null;
        this.notifyListeners();
        this.fetchActivity();
    }

    public setSimulation(activity: Activity): void {
        this.simulation = activity;
        if (this.timer) {
            window.clearTimeout(this.timer);
            this.timer = null;
        }
        this.notifyListeners();
    }

    private fetchActivity(wait: number = 0): void {
        // if (!this.timer) {
        //     this.timer = window.setTimeout(() => {
        //         this.timer = null;
        //         this.getActivityExternal(this.flowUUID)
        //             .then((activity: Activity) => {
        //                 this.activity = activity;
        //                 this.notifyListeners();
        //             })
        //             .catch(() => {
        //                 // ignore missing activity
        //             });
        //         this.fetchActivity(REFRESH_SECONDS * 1000);
        //     }, wait);
        // }
    }

    public notifyListeners(): void {
        // tslint:disable-next-line:forin
        for (const counter in this.listeners) {
            this.listeners[counter].handleRequestUpdate();
        }
    }

    public deregister(key: string): void {
        delete this.listeners[key];
    }

    public registerListener(counter: Counter): void {
        // this is called from ref, which is null on unmounts
        if (counter) {
            this.listeners[counter.getKey()] = counter;
        }
    }

    public getActivity(): Activity {
        if (this.simulation) {
            return this.simulation;
        }
        return this.activity;
    }

    public getActiveCount(nodeUUID: string): number {
        const activity = this.getActivity();
        if (activity) {
            const count = activity.nodes[nodeUUID];
            if (count !== undefined) {
                return count;
            }
        }
        return 0;
    }

    public getPathCount(exit: Exit): number {
        const activity = this.getActivity();
        if (activity) {
            if (exit.destination_uuid) {
                const count = activity.segments[exit.uuid + ':' + exit.destination_uuid];
                if (count !== undefined) {
                    return count;
                }
            }
        }
        return 0;
    }
}
