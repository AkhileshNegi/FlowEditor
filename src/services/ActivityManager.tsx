import { Exit } from '../FlowDefinition';
import { External, Activity } from './External';
import { CounterComp } from "../components/Counter";

// how often we ask the server for new data
const REFRESH_SECONDS = 10;

export class ActivityManager {

    private static singleton: ActivityManager;
    private external: External;
    private activity: Activity;
    private listeners: { [key: string]: CounterComp } = {};

    private timer: any;

    static get(): ActivityManager {
        return ActivityManager.singleton;
    }

    static initialize(external: External) {
        this.singleton = new ActivityManager(external);
    }

    constructor(external: External) {
        this.external = external;
        this.fetchActivity();
        this.registerListener = this.registerListener.bind(this);
    }

    private fetchActivity(wait: number = 0) {
        if (!this.timer) {
            this.timer = window.setTimeout(() => {
                this.timer = null;
                this.external.getActivity().then((activity: Activity) => {
                    this.activity = activity;
                    this.notifyListeners();
                });

                this.fetchActivity(REFRESH_SECONDS * 1000);
            }, wait);
        }
    }

    public notifyListeners() {
        for (let counter in this.listeners) {
            this.listeners[counter].requestUpdate();
        }
    }

    public deregister(key: string) {
        delete this.listeners[key];
    }

    public registerListener(counter: CounterComp) {
        // this is called from ref, which is null on unmounts
        if (counter) {
            this.listeners[counter.getKey()] = counter;
        }
    }

    public getActiveCount(nodeUUID: string): number {
        if (this.activity) {
            var count = this.activity.active[nodeUUID];
            if (count !== undefined) {
                return count;
            }
        }
        return 0;
    }

    public getPathCount(exit: Exit): number {
        if (this.activity) {
            if (exit.destination_node_uuid) {
                var count = this.activity.paths[exit.uuid + ":" + exit.destination_node_uuid];
                if (count !== undefined) {
                    return count;
                }
            }
        }
        return 0;
    }
}
