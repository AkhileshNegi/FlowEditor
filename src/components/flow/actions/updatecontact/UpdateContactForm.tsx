import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import Dialog, { ButtonSet } from '~/components/dialog/Dialog';
import {
    initializeForm,
    sortFieldsAndProperties,
    stateToAction,
    UpdateContactFormState
} from '~/components/flow/actions/updatecontact/helpers';
import {
    ActionFormProps,
    CHANNEL_PROPERTY,
    LANGUAGE_PROPERTY,
    NAME_PROPERTY
} from '~/components/flow/props';
import AssetSelector from '~/components/form/assetselector/AssetSelector';
import TextInputElement from '~/components/form/textinput/TextInputElement';
import TypeList from '~/components/nodeeditor/TypeList';
import { fakePropType } from '~/config/ConfigProvider';
import { Types } from '~/config/typeConfigs';
import { ContactProperties } from '~/flowTypes';
import { Asset, AssetType, updateAssets } from '~/store/flowContext';
import * as mutators from '~/store/mutators';
import { mergeForm } from '~/store/nodeEditor';
import { DispatchWithState, GetState } from '~/store/thunks';
import { validate, validateRequired } from '~/store/validators';
import { createUUID } from '~/utils';

import * as styles from './UpdateContact.scss';

export const CONTACT_PROPERTIES: Asset[] = [NAME_PROPERTY, LANGUAGE_PROPERTY, CHANNEL_PROPERTY];

export default class UpdateContactForm extends React.Component<
    ActionFormProps,
    UpdateContactFormState
> {
    public static contextTypes = {
        assetService: fakePropType
    };

    constructor(props: ActionFormProps) {
        super(props);

        this.state = initializeForm(this.props.nodeSettings);

        bindCallbacks(this, {
            include: [/^get/, /^on/, /^handle/]
        });
    }

    private handleUpdate(keys: {
        type?: Types;
        name?: string;
        channel?: Asset;
        language?: Asset;
        field?: Asset;
        fieldValue?: string;
    }): boolean {
        const updates: Partial<UpdateContactFormState> = {};

        if (keys.hasOwnProperty('type')) {
            updates.type = keys.type;
        }

        if (keys.hasOwnProperty('name')) {
            updates.name = { value: keys.name };
        }

        if (keys.hasOwnProperty('channel')) {
            updates.channel = validate('Channel', keys.channel, [validateRequired]);
        }

        if (keys.hasOwnProperty('language')) {
            updates.language = validate('Language', keys.language, [validateRequired]);
        }

        if (keys.hasOwnProperty('field')) {
            updates.field = { value: keys.field };
        }

        if (keys.hasOwnProperty('fieldValue')) {
            updates.fieldValue = { value: keys.fieldValue };
        }

        const updated = mergeForm(this.state, updates);
        this.setState(updated);
        return updated.valid;
    }

    private handlePropertyChange(selected: Asset[]): boolean {
        const selection = selected[0];
        if (selection.type === AssetType.ContactProperty) {
            switch (selection.id) {
                case ContactProperties.Name:
                    return this.handleUpdate({
                        field: selection,
                        type: Types.set_contact_name,
                        name: ''
                    });
                case ContactProperties.Language:
                    return this.handleUpdate({
                        field: selection,
                        type: Types.set_contact_language
                    });
                case ContactProperties.Channel:
                    return this.handleUpdate({ field: selection, type: Types.set_contact_channel });
            }
        }
        return this.handleUpdate({
            type: Types.set_contact_field,
            field: selection,
            fieldValue: ''
        });
    }

    private handleChannelUpdate(selection: Asset[]): boolean {
        return this.handleUpdate({ channel: selection[0] });
    }

    private handleLanguageUpdate(selection: Asset[]): boolean {
        return this.handleUpdate({ language: selection[0] });
    }

    private handleFieldValueUpdate(fieldValue: string): boolean {
        return this.handleUpdate({ fieldValue });
    }

    private handleNameUpdate(name: string): boolean {
        return this.handleUpdate({ name });
    }

    private onUpdated(dispatch: DispatchWithState, getState: GetState): void {
        const {
            flowContext: { assetStore }
        } = getState();

        if (this.state.field.value.type === AssetType.Field) {
            dispatch(
                updateAssets(mutators.addAssets('fields', assetStore, [this.state.field.value]))
            );
        }
    }

    public handleFieldAdded(name: string): void {
        const newField = { id: createUUID(), name, type: AssetType.Field };
        this.handlePropertyChange([newField]);
    }

    private handleSave(): void {
        let valid = this.state.valid;

        // check if language required
        if (this.state.type === Types.set_contact_language) {
            valid = this.handleLanguageUpdate([this.state.language.value]) && valid;
        }

        // check if channel required
        if (this.state.type === Types.set_contact_channel) {
            valid = this.handleChannelUpdate([this.state.channel.value]) && valid;
        }

        if (valid) {
            // do the saving!
            this.props.updateAction(
                stateToAction(this.props.nodeSettings, this.state),
                this.onUpdated
            );
            this.props.onClose(true);
        }
    }

    private getButtons(): ButtonSet {
        return {
            primary: { name: 'Ok', onClick: this.handleSave, disabled: !this.state.valid },
            secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
        };
    }

    /**
     * The value widget varies for the action type
     */
    private getValueWidget(): JSX.Element {
        if (this.state.type === Types.set_contact_channel) {
            return (
                <AssetSelector
                    key="select_channel"
                    name="Channel"
                    placeholder="Select the channel to use for this contact"
                    assets={this.props.assetStore.channels}
                    entry={this.state.channel}
                    searchable={true}
                    clearable={true}
                    onChange={this.handleChannelUpdate}
                />
            );
        }

        if (this.state.type === Types.set_contact_language) {
            return (
                <AssetSelector
                    key="select_language"
                    name="Language"
                    placeholder="Select the language to use for this contact"
                    assets={this.props.assetStore.languages}
                    entry={this.state.language}
                    searchable={true}
                    clearable={true}
                    onChange={this.handleLanguageUpdate}
                />
            );
        } else if (this.state.type === Types.set_contact_name) {
            return (
                <TextInputElement
                    name="Name"
                    placeholder="Enter a new name for the contact"
                    onChange={this.handleNameUpdate}
                    entry={this.state.name}
                    autocomplete={true}
                    focus={true}
                />
            );
        } else {
            return (
                <TextInputElement
                    name="Field Value"
                    placeholder={`Enter a new value for ${this.state.field.value.name}`}
                    onChange={this.handleFieldValueUpdate}
                    entry={this.state.fieldValue}
                    autocomplete={true}
                    focus={true}
                />
            );
        }
    }

    public render(): JSX.Element {
        const typeConfig = this.props.typeConfig;

        return (
            <Dialog
                title={typeConfig.name}
                headerClass={typeConfig.type}
                buttons={this.getButtons()}
            >
                <TypeList
                    __className=""
                    initialType={typeConfig}
                    onChange={this.props.onTypeChange}
                />

                <p>Select what to update</p>
                <AssetSelector
                    name="Contact Field"
                    assets={this.props.assetStore.fields}
                    additionalOptions={CONTACT_PROPERTIES}
                    entry={this.state.field}
                    searchable={true}
                    sortFunction={sortFieldsAndProperties}
                    onChange={this.handlePropertyChange}
                    onCreateOption={this.handleFieldAdded}
                />

                <div className={styles.value}>{this.getValueWidget()}</div>
            </Dialog>
        );
    }
}
