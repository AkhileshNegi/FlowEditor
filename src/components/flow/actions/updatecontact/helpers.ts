import {
    CHANNEL_PROPERTY,
    LANGUAGE_PROPERTY,
    NAME_PROPERTY,
    UpdateContactFormState
} from '~/components/flow/actions/updatecontact/UpdateContactForm';
import { getTypeConfig, Types } from '~/config/typeConfigs';
import {
    Channel,
    Field,
    Language,
    SetContactAttribute,
    SetContactChannel,
    SetContactField,
    SetContactLanguage,
    SetContactName
} from '~/flowTypes';
import { Asset, AssetType, removeAsset } from '~/services/AssetService';
import { NodeEditorSettings } from '~/store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): UpdateContactFormState => {
    const state: UpdateContactFormState = {
        type: Types.set_contact_name,
        valid: false,
        name: { value: '' },
        channel: { value: null },
        language: { value: null },
        field: { value: NAME_PROPERTY },
        fieldValue: { value: '' }
    };

    if (settings.originalAction) {
        const originalType = settings.originalAction.type;
        // these have aliases, so compare the config we resolve to
        if (getTypeConfig(originalType) === getTypeConfig(Types.set_contact_field)) {
            state.type = originalType;

            switch (originalType) {
                case Types.set_contact_field:
                    const fieldAction = settings.originalAction as SetContactField;
                    state.field = { value: fieldToAsset(fieldAction.field) };
                    state.fieldValue = { value: fieldAction.value };
                    state.valid = true;
                    return state;
                case Types.set_contact_channel:
                    const channelAction = settings.originalAction as SetContactChannel;
                    state.field = { value: CHANNEL_PROPERTY };
                    state.channel = { value: channelToAsset(channelAction.channel) };
                    state.valid = true;
                    return state;
                case Types.set_contact_language:
                    const languageAction = settings.originalAction as SetContactLanguage;
                    state.field = { value: LANGUAGE_PROPERTY };
                    state.valid = true;
                    state.language = {
                        value: languageToAsset({
                            iso: languageAction.language,
                            name: languageAction.language
                        })
                    };
                    return state;
                case Types.set_contact_name:
                    const nameAction = settings.originalAction as SetContactName;
                    state.valid = true;
                    state.name = {
                        value: nameAction.name
                    };
                    return state;
            }
        }
    }

    // default is updating name
    return state;
};

export const stateToAction = (uuid: string, state: UpdateContactFormState): SetContactAttribute => {
    if (state.type === Types.set_contact_field) {
        return {
            uuid,
            type: state.type,
            field: assetToField(state.field.value),
            value: state.fieldValue.value
        };
    }

    if (state.type === Types.set_contact_channel) {
        return {
            uuid,
            type: state.type,
            channel: assetToChannel(state.channel.value)
        };
    }

    if (state.type === Types.set_contact_language) {
        return {
            uuid,
            type: state.type,
            language: assetToLanguage(state.language.value)
        };
    }

    if (state.type === Types.set_contact_name) {
        return {
            uuid,
            type: state.type,
            name: state.name.value
        };
    }
};

export const sortFieldsAndProperties = (a: Asset, b: Asset): number => {
    // Name always goes first
    if (a === NAME_PROPERTY) {
        return -1;
    }

    if (b === NAME_PROPERTY) {
        return 1;
    }

    // go with alpha-sort for everthing else
    if (a.type === b.type) {
        return a.name.localeCompare(b.name);
    }

    // non-name non-fields go last
    if (a.type !== AssetType.Field) {
        return 1;
    }

    if (b.type !== AssetType.Field) {
        return -1;
    }

    return 0;
};

export const fieldToAsset = (field: Field = { key: '', name: '' }): Asset => ({
    id: field.key,
    name: field.name,
    type: AssetType.Field
});

export const assetToField = (asset: Asset): Field => ({
    key: asset.id,
    name: asset.name
});

export const assetToChannel = (asset: Asset): any => {
    if (asset.id === removeAsset.id) {
        return {};
    }

    return {
        uuid: asset.id,
        name: asset.name
    };
};

export const assetToLanguage = (asset: Asset): string => {
    if (asset.id === removeAsset.id) {
        return '';
    }
    return asset.id;
};

export const languageToAsset = ({ iso, name }: Language) => {
    if (!iso || iso.length === 0) {
        return removeAsset;
    }

    return {
        id: iso,
        name,
        type: AssetType.Language
    };
};

export const channelToAsset = ({ uuid, name }: Channel) => {
    if (!uuid) {
        return removeAsset;
    }
    return {
        id: uuid,
        name,
        type: AssetType.Language
    };
};
