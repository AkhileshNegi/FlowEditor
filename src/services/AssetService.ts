// tslint:disable:max-classes-per-file
import { v4 as generateUUID } from 'uuid';
import { Endpoints, FlowEditorConfig, ContactProperties, Group } from '../flowTypes';
import axios, { AxiosResponse } from 'axios';
import { FlowComponents } from '../store/helpers';
import { dump } from '../utils';

export enum AssetType {
    Channel = 'channel',
    Flow = 'flow',
    Group = 'group',
    Property = 'property',
    Field = 'field'
}

export interface Asset {
    id: string;
    name: string;
    type: AssetType;

    isNew?: boolean;
    content?: any;
}

enum IdProperty {
    UUID = 'uuid',
    Key = 'key',
    ID = 'id'
}

enum NameProperty {
    Name = 'name'
}

enum SimAssetType {
    Flow = 'flow',
    Fields = 'field_set',
    Groups = 'group_set',
    Channels = 'channel_set'
}

interface SimAsset {
    type: SimAssetType;
    url: string;
    content: any;
}

export const getBaseURL = (): string => {
    const location = window.location;
    return (
        location.protocol + '//' + location.hostname + (location.port ? ':' + location.port : '')
    );
};

export class Assets {
    public endpoint: string;
    private localStorage: boolean;
    protected idProperty: IdProperty;
    protected assetType: AssetType;
    protected assets: { [id: string]: Asset } = {};

    constructor(endpoint: string, localStorage: boolean) {
        this.localStorage = localStorage;
        this.endpoint = endpoint;
        this.idProperty = IdProperty.UUID;
    }

    public get(id: string): Promise<Asset> {
        const existing = this.assets[id];
        if (existing) {
            return new Promise<Asset>(resolve => {
                resolve(existing);
            });
        }

        return new Promise<Asset>((resolve, reject) => {
            const url = `${this.endpoint}?uuid=${id}`;
            axios
                .get(url)
                .then((response: AxiosResponse) => {
                    const ob = response.data.results[0];
                    const asset = {
                        id: ob.uuid,
                        name: ob.name,
                        type: this.assetType,
                        content: ob.definition
                    };
                    if (this.localStorage) {
                        this.assets[id] = asset;
                    }
                    return resolve(asset);
                })
                .catch(error => reject(error));
        });
    }

    private searchLocalItems(term: string): Asset[] {
        // search our local items first
        const matches: Asset[] = [];
        Object.keys(this.assets).map((key: string) => {
            const asset = this.assets[key];
            if (this.matches(term, asset.name)) {
                matches.push(asset);
            }
        });
        return matches;
    }

    public search(term: string): Promise<Asset[]> {
        const matches = this.searchLocalItems(term);

        // then query against our endpoint to add to that list
        let url = this.endpoint;
        if (term) {
            url += '?query=' + encodeURIComponent(term);
        }

        return axios.get(url).then((response: AxiosResponse) => {
            for (const result of response.data.results) {
                if (this.matches(term, result.name)) {
                    matches.push({
                        name: result.name,
                        id: result[this.idProperty],
                        type: this.assetType
                    });
                }
            }
            return matches;
        });
    }

    public matches(query: string, check: string): boolean {
        if (!check) {
            return false;
        }

        if (query.length === 0) {
            return true;
        }

        return (
            check
                .toLocaleLowerCase()
                .trim()
                .indexOf(query.toLocaleLowerCase().trim()) > -1
        );
    }

    public add(result: Asset): void {
        if (this.localStorage) {
            this.assets[result.id] = result;
        }
    }

    public addAll(results: Asset[]): void {
        results.map((result: Asset) => {
            this.add(result);
        });
    }

    public update(uuid: string, content: any): Promise<Asset> {
        const result = new Promise<Asset>((resolve, reject) => {
            if (this.localStorage) {
                const asset = this.assets[uuid];
                asset.content = content;
                resolve(asset);
            } else {
                const url = `${this.endpoint}?uuid=${uuid}`;
                axios
                    .post(url, content)
                    .then((response: AxiosResponse) => {
                        const ob = response.data.results[0];
                        const asset = {
                            id: ob.uuid,
                            name: ob.name,
                            type: this.assetType,
                            content: ob.definition
                        };
                        if (this.localStorage) {
                            this.assets[uuid] = asset;
                        }
                        return resolve(asset);
                    })
                    .catch(error => reject(error));
            }
        });

        return result;
    }

    public getAssetSet(): any[] {
        const assets: any[] = [];
        Object.keys(this.assets).map((key: string) => {
            const asset = this.assets[key];
            if (asset.type !== AssetType.Property) {
                assets.push({
                    [this.idProperty]: asset.id,
                    name: asset.name,
                    ...asset.content
                });
            }
        });
        return assets;
    }

    public getAssetContents(type: SimAssetType): SimAsset[] {
        const assets: SimAsset[] = [];
        Object.keys(this.assets).map((key: string) => {
            const asset = this.assets[key];
            assets.push({
                type,
                url: getBaseURL() + this.endpoint + '/' + asset.id + '/',
                content: asset.content
            });
        });

        return assets;
    }
}

class GroupAssets extends Assets {
    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.UUID;
        this.assetType = AssetType.Group;
    }
}

class FieldAssets extends Assets {
    public static CONTACT_PROPERTIES: Asset[] = [
        {
            name: ContactProperties.Name,
            id: ContactProperties.Name.toLowerCase(),
            type: AssetType.Property
        }
        /*{ 
            name: ContactProperties.Language, 
            id: ContactProperties.Language.toLowerCase(), 
            type: AttributeType.property 
        }*/
    ];

    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.Key;
        this.assetType = AssetType.Field;

        FieldAssets.CONTACT_PROPERTIES.map((result: Asset) => {
            this.assets[result.id] = result;
        });
    }
}

class ChannelAssets extends Assets {
    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.UUID;
        this.assetType = AssetType.Channel;

        const uuid = '6934255e-446b-40d3-b2ca-c5f801bd2278';
        this.assets[uuid] = {
            id: uuid,
            name: 'Simulator',
            type: AssetType.Channel,
            content: {
                address: '+12065550000',
                schemes: ['tel'],
                roles: ['send', 'receive']
            }
        };
    }
}

class FlowAssets extends Assets {
    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.UUID;
        this.assetType = AssetType.Flow;
    }
}

export default class AssetService {
    private channels: Assets;
    private groups: Assets;
    private fields: Assets;
    private flows: FlowAssets;

    constructor(config: FlowEditorConfig) {
        this.groups = new GroupAssets(config.endpoints.groups, config.localStorage);
        this.fields = new FieldAssets(config.endpoints.fields, config.localStorage);
        this.flows = new FlowAssets(config.endpoints.flows, config.localStorage);

        // channels are always mocked for local
        this.channels = new ChannelAssets('/channels', true);
    }

    public addFlowComponents(flowComponents: FlowComponents): void {
        this.groups.addAll(flowComponents.groups);
        this.fields.addAll(flowComponents.fields);
    }

    public getFlowAssets(): Assets {
        return this.flows;
    }

    public getGroupAssets(): Assets {
        return this.groups;
    }

    public getFieldAssets(): Assets {
        return this.fields;
    }

    public getSimulationAssets(): any {
        let simAssets: SimAsset[] = [];
        const base = getBaseURL();
        // our group set asset
        simAssets.push({
            type: SimAssetType.Groups,
            url: base + this.groups.endpoint + '/',
            content: this.groups.getAssetSet()
        });

        // our fields
        simAssets.push({
            type: SimAssetType.Fields,
            url: base + this.fields.endpoint + '/',
            content: this.fields.getAssetSet()
        });

        // our channels
        simAssets.push({
            type: SimAssetType.Channels,
            url: base + this.channels.endpoint + '/',
            content: this.channels.getAssetSet()
        });

        // our flows
        simAssets = simAssets.concat(this.flows.getAssetContents(SimAssetType.Flow));

        const payload = {
            assets: simAssets,
            asset_server: {
                type_urls: {
                    flow: base + this.flows.endpoint + '/{uuid}/',
                    field_set: base + this.fields.endpoint + '/',
                    channel_set: base + this.channels.endpoint + '/',
                    group_set: base + this.groups.endpoint + '/'
                }
            }
        };

        return payload;
    }
}
