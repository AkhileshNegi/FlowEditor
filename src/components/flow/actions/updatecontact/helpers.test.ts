import {
    assetToChannel,
    assetToLanguage,
    channelToAsset,
    fieldToAsset,
    languageToAsset,
    sortFieldsAndProperties
} from '~/components/flow/actions/updatecontact/helpers';
import { CONTACT_PROPERTIES } from '~/components/flow/actions/updatecontact/UpdateContactForm';
import { removeAsset } from '~/services/AssetService';

describe('UpdateContact.helpers', () => {
    it('should sort options', () => {
        const fields = require('~/test/assets/fields.json');
        for (const field of fields.results) {
            CONTACT_PROPERTIES.push(fieldToAsset(field));
        }
        expect(CONTACT_PROPERTIES.sort(sortFieldsAndProperties)).toMatchSnapshot();
    });

    it('should return remove asset', () => {
        expect(assetToChannel(removeAsset)).toEqual({});
        expect(assetToLanguage(removeAsset)).toEqual('');

        expect(fieldToAsset()).toEqual({ id: '', name: '', type: 'field' });
        expect(fieldToAsset({ key: 'gender', name: 'Gender' })).toEqual({
            id: 'gender',
            name: 'Gender',
            type: 'field'
        });
        expect(languageToAsset({ iso: '', name: '' })).toEqual(removeAsset);
        expect(channelToAsset({ uuid: '', name: '' })).toEqual(removeAsset);
    });
});
