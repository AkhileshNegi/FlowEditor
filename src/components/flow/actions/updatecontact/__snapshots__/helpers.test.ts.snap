// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateContact.helpers should sort options 1`] = `
Array [
  Object {
    "id": "name",
    "name": "Name",
    "type": "property",
  },
  Object {
    "id": "expected_delivery_date",
    "name": "Expected Delivery Date",
    "type": "field",
  },
  Object {
    "id": "national_id",
    "name": "National ID",
    "type": "field",
  },
  Object {
    "id": "number_of_children",
    "name": "Number of Children",
    "type": "field",
  },
  Object {
    "id": "channel",
    "name": "Channel",
    "type": "property",
  },
  Object {
    "id": "language",
    "name": "Language",
    "type": "property",
  },
]
`;
